### PR TITLE
Avoid forwarding dns queries of type DS to recursors (#6183)

### LINKED ISSUE #6183: Consul forwards DNS queries inside consul. zone to recursors for unknown record types
#### Overview of the Issue

When consul is configured with a recursor, DNS queries for unrecognized record types are forwarded to the configured recursors. This forwarding also happens when resolving names in the `consul.` zone with types other than `SOA`, `NS`, `ANY`, `A`, `AAAA`, `TXT`.

In a particular setup, when dnsmasq is configured to forward `consul.` to a consul DNS server, and consul is configured to use this dnsmasq server as a recursor, this results in a DNS recursion loop. The dnsmasq configuration is described on hashicorp learn: https://learn.hashicorp.com/consul/security-networking/forwarding#dnsmasq-setup

The documentation states that consul only forwards DNS queries for names outside the `consul.` zone: https://www.consul.io/docs/agent/options.html#recursors

My expectations would also be that consul acts as an authoritative DNS server for the `consul.` zone, and does not forward it to recursors. 

#### Reproduction Steps

A docker setup to reproduce this issue is available in a gist: https://gist.github.com/vierbergenlars/d5877cf8bb076fb5789f47d1ad7039fb

Alternatively:

1. Set up a consul node (or a cluster, it does not matter) with `-recursor` pointing to dnsmasq.
1. Set up dnsmasq, forwarding the `consul.` zone to the consul DNS server.
1. Perform a DNS lookup for an unsupported record type to `consul.service.consul` against either dnsmasq or consul. I used the dnssec `DS` type. `dig @127.0.0.1 -p 8600` DS consul.service.consul`
1. Even long after `dig` has finished, DNS requests keep bouncing around between consul and dnsmasq.

### Consul info for Server

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = a82e6a7f
	version = 1.5.2
consul:
	acl = disabled
	bootstrap = true
	known_datacenters = 1
	leader = true
	leader_addr = 127.0.0.1:8300
	server = true
raft:
	applied_index = 71
	commit_index = 71
	fsm_pending = 0
	last_contact = 0
	last_log_index = 71
	last_log_term = 2
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:a1a793cb-2564-e6ab-8c85-6bc4a3ee35b1 Address:127.0.0.1:8300}]
	latest_configuration_index = 1
	num_peers = 0
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 2
runtime:
	arch = amd64
	cpu_count = 4
	goroutines = 82
	max_procs = 4
	os = linux
	version = go1.12.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 1
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
```

</details>

### Operating system and Environment details

OS, Architecture, and any other information you can provide about the environment.

### Log Fragments

From the consul server:

```
    2019/07/20 16:08:42 [ERR] dns: recurse failed: read udp 10.33.10.2:34942->10.33.10.3:53: i/o timeout
    2019/07/20 16:08:42 [ERR] dns: all resolvers failed for {consul.service.consul. 43 1} from client 10.33.10.3:571 (udp)
    2019/07/20 16:08:42 [DEBUG] dns: request for {consul.service.consul. 43 1} (udp) (2.000994556s) from client 10.33.10.3:571 (udp)
    2019/07/20 16:08:42 [ERR] dns: recurse failed: read udp 10.33.10.2:59383->10.33.10.3:53: i/o timeout
    2019/07/20 16:08:42 [ERR] dns: all resolvers failed for {consul.service.consul. 43 1} from client 10.33.10.3:46076 (udp)
    2019/07/20 16:08:42 [ERR] dns: recurse failed: read udp 10.33.10.2:36895->10.33.10.3:53: i/o timeout
    2019/07/20 16:08:42 [ERR] dns: all resolvers failed for {consul.service.consul. 43 1} from client 10.33.10.3:33552 (udp)
    2019/07/20 16:08:42 [DEBUG] dns: request for {consul.service.consul. 43 1} (udp) (2.001374786s) from client 10.33.10.3:46076 (udp)
    2019/07/20 16:08:42 [DEBUG] dns: request for {consul.service.consul. 43 1} (udp) (2.002385645s) from client 10.33.10.3:33552 (udp)
```

From dnsmasq:

```
dnsmasq: Maximum number of concurrent DNS queries reached (max: 150)
dnsmasq: Maximum number of concurrent DNS queries reached (max: 150)
```

### PR DESCRIPTION
**Pull request by vierbergenlars**
Answer all `DS` queries that are sent for our domains to recursors with a refused response code.

The upstream servers will not have a `DS` record for consul domains anyways, because Consul does not support DNSSEC yet, but also because the domain will not be known by the resolvers.

Fixes #6183

### CODE REVIEW

**Comment by i0rek** [comment]
Thank you @vierbergenlars for your work. Is this still something you would like to add?

**Comment by vierbergenlars** [comment]
Yes, this is still something I would like to add

**Comment by i0rek** [comment]
Thanks for your reply! This sounds like a nice patch and I would like to ask you to rebase it so that I can review.

**Comment by vierbergenlars** [comment]
Rebased on top of master.

Note that the diff is best viewed with whitespace changed disabled, because I added a condition and had to indent a large block of code.

File: agent/dns.go [hunk 1]
```diff
@@ -1760,56 +1760,79 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
...
-				"rtt", rtt,
-				"recursor", recursor,
-			)
-			if err := resp.WriteMsg(r); err != nil {
-				d.logger.Warn("failed to respond", "error", err)
-			}
-			return
+	// Refuse to recursively resolve requests of type DS for domains that we manage.
+	// Upstream resolvers will not know about the correct record anyways, and it might result in a resolver loop
+	if q.Qtype == dns.TypeDS {
```
Comment (i0rek) [inline agent/dns.go]: Could you extract a func for this block? Eg `func (d *DNSServer) refuseRecurseDS(req *dns.Msg, domain string) bool`.

And also a func for creating the message `func generateMessage(code int)`.

Afterwards we can write:

```golang
for _, domain := range []string{d.domain, d.altDomain} {
  if d.refuseRecurseDS(req, domain) {
    d.logger.Debug("request for %v (%s): Refusing to recurse type DS requests to %v.", q, network, domain)
    m := generateMessage(dns.RcodeRefused)
    resp.WriteMsg(m)
    return
  }
}
```

That would also get rid of the big indentation below. And `generateMessage` could be reused for the `dns.RcodeServerFailure` message at the end.

**Pull request closed without merging**